README.md
# ÜberSite

An internal website designed for use on [&Uuml;bertweak](http://ubertweak.org.au) and [related camps](http://crutech.org.au). 

## Requirements

* A web server with `mod_rewrite` (or compatible extension) enabled. Apache 2.2 or above is highly recommended.
* A MySQL server.
* PHP 5.4 (may work on 5.3, but I wouldn't count on it).
* The [MySQLi](http://php.net/mysqli) extension for PHP must be enabled.

#### Non-core requirements

* [GD](http://php.net/manual/en/book.image.php) or [ImageMagick](http://php.net/imagick) is required for thumbnail generation.
* [Exif](http://php.net/exif) is required for photo uploads.
* [SSH2](http://php.net/ssh2) is required if you want to use SSH authentication.
* [LDAP](http://php.net/ldap) is required if you want to use LDAP authentication.

## Installation

Very little setup is required; just drag the files into a directory of your choice and go. Note that at this time, from the web browser perspective, the files must be in the domain root. Whether you accomplish this by setting by a vhost or simply placing the files in the webserver root is up to you.

Once you've copied the files, you should be able to access the website in your browser. You'll have to go through a brief setup process which isn't overly complicated.

By default, &Uuml;berSite has developer mode enabled. You will want to disable this at some point by going into `camp-data/config/config.json` and setting `developerMode` to `false`.

Chances are you'll want to change what items show up in the menu bar. You can do this by going into `camp-data/config/menu.json` and changing the `*requirements` field for the relevant menu items to `false`.

### That's not very intuitive

Welcome to &Uuml;berSite, enjoy your stay. We're working on making things easier, but it's a slow process.

## Things to be aware of

### Authentication types
SSH authentication barely works, and LDAP authentication doesn't really work at all. I really wouldn't recommend using them. Just stick with MySQL. Really.

### Timetable data

By default, there is no timetable data entered into the database. There's an example timetable located in `setup/crutech_timetable.sql` that you can import to see what the data should look like. There's no built in editor for the timetable yet, so you'll have to hop into the database and edit the data yourself if you want to change things.

### Trosnoth achievements

The process for importing Trosnoth achievements is incredibly complicated. One day we'll document the process, but until then, it's recommended not to bother unless you know what you're doing.

### Questionnaire

The questionnaire is a fickle beast, more prone to breaking than not. It's also in the middle of a large refactor, so it's neither here nor there at this point. My advice: wait for the documentation.

### Documentation

In general, things aren't very well documented (have you noticed?), so if something fails to "just work", that you may be in a spot of bother. Thankfully, the simpler things such as quotes, photos and polls have been pretty well tested, so you shouldn't have too many issues.

docs/changelog.md
# ÜberSite Changelog-

## ÜberSite Fenix `current development version`

### v2.1.0:
* Full tidy URL support for all pages
* Added a proper readme!
* Replaced "Trac" link in developer mode with "Github" link
* Converted changelog to markdown format and reversed the chronological order (newest changes are now at the top)
* Fixed case-sensitivity issue which caused thumbnails in the photo processor to sometimes not show up
* Introduced Twig support to replace the extremely-outdated bTemplate
* Introduced DBV for better database versioning
* #19: The new PHP password hashing API is now used throughout the code instead of the old method
* #16: Added a CLI script to reset a user's password
* Added a basic User class
* Major changes to the way alert/success messages are displayed, some some visual adjustments
* Replaced version.json with a static Software class
* Added Code Challenge functionality

## ÜberSite Executor (All 2012 camps)

### v2.0.0 (Changes made on Crütech):
* **Moved files to a public Github repository**
* Changed the internal directory structure around into more logical components
* Added tidy URLs for some parts of the website
* Reduced the number of row-spans the timetable uses
* Added questionnaire writer and changed how the questionnaire gets its questions
* Basic support for incremental database upgrades
* Added an über button
* Photo processing improvements

### v2.0.0 (Changes made up to Spring):
* Partial code restructure in an attempt to conform to some sort of consistant standard
* Added options for multiple choice polls and polls with hidden results
* Minor changes to the setup process
* Questionnaire feedback changes that made it easier to view on non-screen mediums

## ÜberSite Daggoth (Spring 2011)

### v1.3.0:
* Minor bug fixes only

## ÜberSite Confederacy (Winter and Nanobyte 2011)

### v1.2.2 (Changes made on Nanobyte):
* Added a textbox for Awards
* Removed Trosnoth Achievement and Website Suggestions for Nanobyte

### v1.2.1 (Changes made on Winter):
* Implemented very basic support for LDAP and SSH authentication
* Added support for auto log-ons
* trosnoth-check.php now uses the proper authentication wrappers
* Removed campers from Pegasaurus page
* Fixed poll moderation not showing up when there weren't any polls already
* Questionnaire feedback can now be shown within Study Groups
* Added a "Bug Report" option for Trosnoth suggestions for Josh to use
* You can now see the rarest achievement that each user has on the Trosnoth page

### v1.2.0:

#### NEW FEATURES FOR CAMPERS:
* New profile page
* New tagging system
* Improved quotes page
* Award nominations
* Pegasaurus

#### NEW FEATURES FOR LEADERS:
* New file upload system

#### OTHER CHANGES:
* #20: Users are now logged out after 15 minutes of inactivity.
* Suggestions are no longer permanently deleted.
* Changelog removed from System Information
* phpThumb is now used as an object.
* Updated filenames and database tables to be somewhat consistant.
* #43: All user input is now properly sanitized before output.
* MySpace completely removed from contacts page, replaced with Facebook.
* #10: Fixed bug that would cause AJAX box to break
* #47: bTemplate if-statement bug finally fixed.
 * Introduced !if statement
* #52: "Standalone mode" implemented which allows direct saving of pages that
  require no external resources.
* #56: Create a Page View Analysis page which shows which pages are most popular.
* Database errors now have a proper error page.
* #34: Elective questions on the Questionnaire (Stage 3) are now pulled from
  the database. A utility page was added to ensure that the tables stay in sync.
* #57: Added support for alternate authentication methods (by using wrappers for
  account-related functions).
* It is now impossible to delete your own account from the accounts page.
* Post-processing process for templates are now standardised.
* Number of MySQL queries used by pages is now significantly smaller.
* voting.php is now accessible by new users.
* Navigation menu is now generated at runtime.
 * #56: Small resolutions now use drop-down menus.
* #60: Profile pages have now been completely revamped with a better look and a
  number of new features.
* #58: Questionnaire questions are now colour-coded in order to remove any
  possible ambiguity in the responses.
* #53: Added navigation buttons when viewing individual photos.
* #18: Users can now be tagged in photos.
* #46: Campers are now able to submit polls for moderation.
* Any quotes, captions or polls that leaders submit are now automatically
  approved.
* #62: photo_view.php will now display correctly at 1024x768.
* #31: Quote page code is now significantly neater.
* #4: A dropdown list is now provided when creating a new quote to allow users
  to select the person they are quoting: this can then be linked back to their
  profile.
* #30: Leaders can now upload photos directly to the website.
* Created the Photo Processing Lab which allows admins to sort and categorise
  uploaded photos. (Currently disabled)
* #64: Profile pages have been completely redesigned
* #65: Added a constant for Nanobyte which removes some links from the navbar
  as well as changing a few other things
* #21: Layout-based CSS and colour-based CSS are now in separate files to allow
  for easier theming
* #74: Added a page for Awards for Nanobyte
* #76: Added a page for Pegasaurus for Winter

## ÜberSite Braxis (Spring 2010):

### v1.1.1:
* Resetting passwords actually works now.
* Added (very basic) support for activity groups.
* Updated Questionnaire and feedback page.
* Added Trosnoth Achievements page and related helper pages for importing data.
* Fixed thumbnail generation on Linux.
* addCampers.php now provides an input form for data.
* Fixed bug in photos.php that wouldn't show the "no photos" message when there
  weren't any photos.
* Usernames are now automatically converted to lowercase on login.
* Recent Changes link no longer shows up in wget mode.

includes/start.php
<?php
  # Set some default values and include some files
  error_reporting(E_ALL);
  ini_set('display_errors', 'On');
  if (!file_exists("camp-data/config/config.json")) {
    header("Location: /setup/setup.php");
  }

  require_once("libraries/bTemplate.php");
  $tpl = new bTemplate();

  include_once("jsonLoader.php");

  date_default_timezone_set($TIMEZONE);

  include_once("database.php");

  $tpl->set("campname", $CAMP_NAME);
  $tpl->set("campyear", $CAMP_YEAR);
  $tpl->set("stylesheet", $STYLESHEET);

  $messages = new MessageQueue();

  $user = new NullUser();
  $feedback = false;
  $questionnaire = false;
  $screenWidth = 1024;
  $idleTime = 60 * 15; // 15 minutes
  $script = explode("/", $_SERVER['SCRIPT_NAME']);
  $pageName = $PAGE;

  $queryCount = 0;

  # Stupid magic quote workaround
  if (get_magic_quotes_gpc()) {
    foreach ($_POST as $key => $value) {
      $_POST[$key] = stripslashes($value);
    }
  }

  # Some helper functions
  include_once("functions.php");

  # Process user session and details
  session_start();

  if (isset($_SESSION['loggedout'])) {
    $messages->addMessage(new Message("warning",
      "You were idle for too long and were automatically logged out."));
    unset($_SESSION['loggedout']);
  }

  if (isset($_SESSION['message'])) {
    $_msgType = $_SESSION['message'][0];
    $_msgText = $_SESSION['message'][1];
    if (isset($_SESSION['message'][2])) {
      $storedValue = $_SESSION['message'][2];
    }

    $messages->addMessage(new Message($_msgType, $_msgText));
    unset($_SESSION['message']);
  }

  # Populate array of users (UserID => Name)
  $query = "SELECT * FROM `people`";
  $result = do_query($query);
  $people = array();
  while ($row = fetch_row($result)) {
    $oneUser = new User($row);
    $people[$row['UserID']] = $oneUser;
  }

  $query = "SELECT `UserID`, `Group` FROM `people_groups`";
  $result = do_query($query);
  $groups = array();
  while ($row = fetch_row($result)) {
    if (!isset($groups[$row['UserID']])) {
      $groups[$row['UserID']] = array();
    }
    $groups[$row['UserID']][] = $row['Group'];
  }

  if (substr($_SERVER['HTTP_USER_AGENT'], 0, 5) == "Wget/") {
    $wget = true;
    $_SESSION['username'] = "wget";
    $_SESSION['wget'] = true;
  } else {
    if (isset($_SESSION['wget'])) {
      echo "You are coming out of wget mode. Please follow through to the <a href='/logout'>logout</a> page.";
      die();
    }
    $wget = false;
    if (isset($_SESSION['username'])) {
      $username = $_SESSION['username'];
      // If the logged in user no longer exists, something bad happened.
      if (!isset($people[$username])) {
        header("Location: /logout");
      }
      $user = $people[$username];

    } else {
      # Redirect to login page if not logged in
      if ($pageName != "login") {
        if ($pageName == "logout") {
          header("Location: /login");
        } else {
          header("Location: /login/$pageName");
        }
      }
    }
  }
  $curTime = time();

  # Disable error reporting for non-admins
  if (!$user->isAdmin() && !$DEVELOPER_MODE) {
     error_reporting(0);
  }

  # If the last page load was more than 15 minutes ago, log the user out
  if (isset($_SESSION['time']) && !$DEVELOPER_MODE) {
    $difference = $curTime - $_SESSION['time'];
    if ($difference > $idleTime && $user->LoggedIn && !$wget && $AUTHENTCATION_TYPE != "ssh") {
      session_destroy();
      session_start();

      # If it was more than 6 hours ago they probably don't care any more.
      if ($difference > 60*60*6) {
        $_SESSION['loggedout'] = true;
      }
      header("Location: /login");
    }
  }

  # Find out what's currently happening
  $_SESSION['time'] = $curTime;

  if (!$wget) {
    $activity = whats_on();
    $tpl->set('whatson', $activity);
    if (isset($activity['Activity'])) {
      $tpl->set('enabled', true, true);
    } else {
      $tpl->set('enabled', false, true);
    }
  }

  # FRIDAY, FRIDAY, GOTTA GET DOWN ON FRIDAY
  # EVERYBODY'S LOOKING FORWARD TO THE WEEKEND
  $TGIF = false;

  $day = date("D");
  $hour = date("G");

  if (date("D") == "Fri" || isset($_SESSION['friday'])) {
    $TGIF = true;
    $day = "Fri";
  }

  eval('$questionnaire = ' . $QUESTIONNAIRE_CONDITION . ';');

  if (!isset($_SESSION['screenWidth']) && !isset($_GET['screenWidth'])) {
    $tpl->set('processWidth', true, true);
  } else {
    if (isset($_GET['screenWidth'])) {
      $width = $_GET['screenWidth'];
      if (is_numeric($width) && (int)$width == $width) {
        $_SESSION['screenWidth'] = $width;
      } else {
        // If the input is invalid, assume worst-case scenario.
        $_SESSION['screenWidth'] = 1024;
      }
    }

    // The screen width should never be below 1024
    $screenWidth = max(1024, $_SESSION['screenWidth']);
    $tpl->set('processWidth', false, true);
  }

  if ($user->LoggedIn && !$wget) {

    # Check if the user needs to change their password
    if ($AUTH_TYPE == "mysql" && $user->needsPasswordChange()) {
      if ($pageName != "change-password") {
        header("Location: /change-password");
      }
    }

    # Check if there are any captions that need to be approved
    if (($pageName != "photos") and ($pageName != "view-photo") and ($user->isAdmin())) {
      $count = checkCaptions();
      if ($count > 0) {
        $messages->addMessage(new Message("alert",
          "You have $count <a href='/photos/admin'>unapproved photo ".suffix("caption", $count) .
          "</a> to approve or destroy."));
      }
    }

    # Check if there are any quotes that need to be approved
    if (($pageName != "quotes") and ($user->isAdmin())) {
      $query = "SELECT COUNT(*) FROM `quotes` WHERE `Status` = 0";
      $result = do_query($query);
      $row = fetch_row($result);
      if ($row[0] > 0) {
        $messages->addMessage(new Message("alert",
          "You have {$row[0]} <a href='/quotes'>unapproved ".suffix("quote", $row[0]) .
          "</a> to approve or destroy."));
      }
    }

    # Check if there are any polls that need to be approve
    if ($pageName != "polls" and $user->isAdmin()) {
      $query = "SELECT COUNT(*) FROM `poll_questions` WHERE `Status` = 0";
      $result = do_query($query);
      $row = fetch_row($result);
      if ($row[0] > 0) {
        $messages->addMessage(new Message("alert",
          "You have {$row[0]} <a href='/polls'>unapproved ".suffix("poll", $row[0]) .
          "</a> to approve or destroy."));
      }
    }

    //if (($pageName != "photo_processing") and ($user->isAdmin()) and (!$DEVELOPER_MODE)) {
    //  $query = "SELECT COUNT(*) FROM `photo_processing` WHERE `Reviewer` IS NULL";
    //  $result = do_query($query);
    //  $row = fetch_row($result);
    //  if ($row[0] > 0) {
    //    $messages->addMessage(new Message("alert",
    //      "There are {$row[0]} <a href='/photo_processing'>unreviewed ".suffix("photo", $row[0]) .
    //      "</a> that need processing."));
    //  }
    //}

    # Check if the user has filled in their profile
    if (!isset($_GET['id'])) {
      $_GET['id'] = false;
    }

    if (($pageName != "person") and ($_GET['id'] != $user->UserID)) {
      if ($TGIF && $CONTACT_DETAILS) {
        $query = "SELECT COUNT(*) FROM `contacts` WHERE `UserID` = '{$user->UserID}'";
        $result = do_query($query);
        $row = fetch_row($result);
        if ($row[0] === '0') {
          $messages->addMessage(new Message("alert",
            "Don't forget to fill in your <a href='/person/$username'>contact information</a> " .
            "if you want to keep in touch with other guys on camp! (Don't click the link if you " .
            "are in the middle of the questionnaire though!)"));
        }
      }
      if (!$TGIF) {
        if (!$user->HasProfile) {
          # Intro games check
          if (date("D") == "Sun") {
            $_SESSION['message'] = array("warning", "Hold it! You need to fill in your profile before you can go any further.");
            header("Location: /person/$username");
          }
          $messages->addMessage(new Message("alert",
            "You have not yet filled in your <a href='/person/$username'>about page</a>: " .
            "why not do that now?"));
        }
      }
    }
    if ($_GET['id'] === false) {
      unset($_GET['id']);
    }

    if ($DEVELOPER_MODE && $user->isAdmin()) {
      $questionnaire = true;
    }

    # Check if the user has submitted their questionnaire
    if ($pageName != $QUESTIONNAIRE_PAGE && $questionnaire && $user->isCamper()) {
      if (num_rows(do_query("SELECT `UserID` FROM `questionnaire` WHERE `UserID` = '$username'")) === 0) {
        $messages->addMessage(new Message("alert",
          "You have not yet filled in your <a href='$QUESTIONNAIRE_PAGE'>questionnaire</a>. " .
          "Please fill it out now."));
        $questionnaire = true;
      }
    }

    # Check if there has been any feedback submitted for the questionnaire yet
    if ($user->isLeader() && num_rows(do_query("SELECT `UserID` FROM `questionnaire` INNER JOIN `people` USING(`UserID`) WHERE `category` = 'camper'"))) {
      $feedback = true;
    }

    /*if ($user->isAdmin()) {
      $feedback = true;
    }*/

    # Record the page access
    $page = $PAGE;
    $reqString = userInput($_SERVER['REQUEST_URI']);
    $userAgent = userInput($_SERVER['HTTP_USER_AGENT']);
    $filename = userInput($_SERVER['SCRIPT_FILENAME']);
    if (isset($_SERVER['HTTP_REFERER'])) {
      $refer = "'".userInput($_SERVER['HTTP_REFERER'])."'";
    } else {
      $refer = "NULL";
    }

    # Don't bother recording automatic page refreshes.
    if (substr($reqString, -11) != "autorefresh") {
      $query = "INSERT INTO `access` (`Timestamp`, `UserID`, `Page`, `RequestString`, `RequestMethod`, `IP`, `UserAgent`, `Filename`, `Refer`) ";
      $query .= "VALUES (NOW(), '$username', '$page', '$reqString', '{$_SERVER['REQUEST_METHOD']}', '{$_SERVER['REMOTE_ADDR']}',";
      $query .= "'$userAgent', '$filename', $refer)";
      do_query($query);
    }

  }

  if (isset($_GET['standalone'])) {
    $tpl->set('standalone', false, true);
    $wget = true;
    $standalone = true;
    $tpl->set('standalone-logo', dataURI("resources/img/logo.png", "image/png"));
    $tpl->set('stand-alone-icon', dataURI("resources/img/icon.png", "image/png"));
    $layoutCSS = file_get_contents("resources/css/layout.css");
    $colourCSS = file_get_contents("resources/css/$STYLESHEET.css");
    $tpl->set('standalone-style', $layoutCSS . "\n\n" . $colourCSS);
  } else {
    $tpl->set('standalone', true, true);
    $standalone = false;
  }

  # If the position is equal to an existing position it will be made into a child.
  # If the parent is not specified, it will be put under "Other Stuff".
  # Items are added in order of priority (zero is the best priority).

  $menu = json_decode(file_get_contents("camp-data/config/menu.json"), true);

  $pages = array();
  foreach ($menu as $filename => $item) {
    $itemArray = array(
        "name" => false,
        "pixels" => 0,
        "priority" => 0,
        "special" => false,
        "longName" => false,
        "parent" => false,
        "requirements" => true);

    if (isset($item["requirements"])) {
      eval('$item["requirements"] = ' . $item["requirements"] . ';');
    }

    if (!isset($item["longName"])) {
      $item["longName"] = $item["name"];
    }
    $pages[$filename] = array_merge($itemArray, $item);
  }

  // Magic numbers!
  $availablePixels = $screenWidth - 17 - 10 - 65 - 226;
  if ($DEVELOPER_MODE) {
    $availablePixels -= 45;
  }

  // Step 1: get rid of all the pages we aren't going to use.
  foreach ($pages as $pageName => $information) {
    if (!$information["requirements"]) {
      unset($pages[$pageName]);
    }
  }

  $menu = array();
  $children = array();
  $couldNotAdd = array();
  $additionLog = array();

  // Step 2: loop through each priority, adding items as we go.
  $priority = 0;
  while ($priority <= 9) {
    foreach ($pages as $pageName => $information) {
      if ($pageName == "other-stuff") {
        continue; // Special case
      }
      if ($information["priority"] === $priority) {
        // Check if this item should be a child
        if (isset($menu[$information["position"]])) {
          $children[$information["position"]][] = $pageName;

        // Check if there are enough pixels for this item to fit
        } else if ($information["pixels"] <= $availablePixels &&
                   $information["name"]) {
          $menu[$information["position"]] = $pageName;
          $availablePixels -= $information["pixels"];
          $additionLog[] = $information["position"];

        // Oh noes! Out of room!
        } else {
          $couldNotAdd[] = $pageName;
        }
      }
    }
    $priority += 1;
  }

  // Step 3: add any items that couldn't fit as children.
  if (!empty($couldNotAdd)) {
    // Add the Other Stuff menu item
    $information = $pages["other-stuff"];
    $menu[$information["position"]] = "other-stuff";
    $availablePixels -= $information["pixels"];

    // This loop will remove items in the reverse order we added them.
    while ($availablePixels < 0) {
      $mostRecent = array_pop($additionLog);
      $couldNotAdd[] = $menu[$mostRecent];
      $information = $pages[$menu[$mostRecent]];
      $availablePixels += $information["pixels"];
      unset($menu[$mostRecent]);
    }

    // Add the left-over items as children
    foreach ($couldNotAdd as $loner) {
      $parent = $pages[$loner]["parent"];
      if (!$parent) {
        $parent = 'other-stuff';
      }
      $parentPos = $pages[$parent]["position"];
      $children[$parentPos][$pages[$loner]["position"]] = $loner;
      ksort($children[$parentPos]);
    }
  }

  // Sort the menu according to position
  ksort($menu);

  if ($user->LoggedIn || $wget) {
    $loginURL = "";
  } else {
    $loginURL = "/login";
  }

  // Step 4: construct the HTML for the navigation bar.
  $menuHTML = "";
  foreach ($menu as $key => $filename) {
    // Check if it's a special item
    if ($pages[$filename]["special"]) {
      $menuHTML .= "<li class=\"special\">";
    } else {
      $menuHTML .= "<li>";
    }

    // We don't want the Other Stuff button to link anywhere
    if ($filename == "other-stuff") {
      $menuHTML .= "<a>{$pages[$filename]['name']}</a>";
    } else {
      $menuHTML .= "<a href='{$loginURL}/{$filename}'>{$pages[$filename]['name']}</a>";
    }

    // Check if the item has any children
    if (isset($children[$key])) {
      $menuHTML .= "\n\t<ul>\n";
      foreach ($children[$key] as $childName) {
        $menuHTML .= "\t<li><a href='{$loginURL}/{$childName}'>{$pages[$childName]['longName']}</a></li>\n";
      }
      $menuHTML .= "\t</ul>";
    }
    $menuHTML .= "</li>\n";
  }

  if ($screenWidth > 1024 || $wget || !$User->LoggedIn) {
    $small = false;
  } else {
    $small = true;
  }

  $tpl->set('menu', $menuHTML);
  $tpl->set('small', $small);

  $tpl->set('head', false);

  $tpl->set('whatson', $tpl->fetch('templates/whats-on.tpl'));
  $tpl->set('shortTitle', "");

  $tpl->set('js', false);
  $tpl->set('loggedin', $user->LoggedIn);
  $tpl->set('loginURL', $loginURL);

  $tpl->set('questionnaire', $questionnaire);
  $tpl->set('feedback', $feedback);

  $tpl->set('currentUser', $username);
  if ($username) {
     $tpl->set('currentName', $people[$username]);
  }

  $tpl->set('config', $jsonConfig);

  $tpl->set('leader', $user->isLeader());
  $tpl->set('camper', $user->isCamper());
  $tpl->set('admin', $user->isAdmin());
  $tpl->set('developer', $DEVELOPER_MODE);

  $tpl->set('wget', $wget);

  $tpl->set('showQueries', $SHOW_QUERIES);

  // TODO: remove this when possible
  $tpl->set('softwareFullName', Software::getFullName());

  // New stuff! Part of the 2012 refactor
  // TODO: we probably shouldn't be using $twig->addGlobal so much
  $twig->addGlobal("user", $user);
  $tpl->set("messages", $messages->getAllMessageHTML());
  if ($_SERVER["REQUEST_METHOD"] == "POST") {
    $twig->addGlobal("form", $_POST);
  }
  $twig->addGlobal("software", new Software());
?>